src/legacy/core_plugins/kibana/public/dashboard/migrations/types.ts

```diff
@@ -26,13 +27,107 @@ export interface SavedObjectAttributes {
 }
 
 interface DashboardAttributes extends SavedObjectAttributes {
+  panelsJSON: string;
+  description: string;
+  version: number;
+  timeRestore: boolean;
+  useMargins: boolean;
+}
+
+export type DashboardAttributes730ToLatest = DashboardAttributes;
+
+interface DashboardAttributesTo720 extends SavedObjectAttributes {
   panelsJSON: string;
   description: string;
   uiStateJSON: string;
   version: number;
   timeRestore: boolean;
+  useMargins: boolean;
 }
 
-export type DashboardDoc = Doc<DashboardAttributes>;
+export type DashboardDoc730ToLatest = Doc<DashboardAttributes>;
+
+export type DashboardDoc700To720 = Doc<DashboardAttributesTo720>;
+
+export type DashboardDocPre700 = DocPre700<DashboardAttributesTo720>;
+
+// Note that these types are prefixed with `Raw` because there are some post processing steps
+// that happen before the saved objects even reach the client. Namely, injecting type and id
+// parameters back into the panels, where the raw saved objects actually have them stored elsewhere.
+//
+// Ideally, everywhere in the dashboard code would use references at the top level instead of
+// embedded in the panels. The reason this is stored at the top level is so the references can be uniformly
+// updated across all saved object types that have references.
 
-export type DashboardDocPre700 = DocPre700<DashboardAttributes>;
+// Starting in 7.3 we introduced the possibility of embeddables existing without an id
+// parameter.  If there was no id, then type remains on the panel.  So it either will have a name,
+// or a type property.
+export type RawSavedDashboardPanel730ToLatest = Pick<
+  RawSavedDashboardPanel640To720,
+  Exclude<keyof RawSavedDashboardPanel640To720, 'name'>
+> & {
+  // Should be either type, and not name (not backed by a saved object), or name but not type (backed by a
+  // saved object and type and id are stored on references).  Had trouble with oring the two types
```

Comment:
"oring" is a typo. Not sure what you meant, maybe "storing"?

Reply:
haha, I couldn't figure out what I meant at first too... then I remembered...  OR'ing.